File: backend/python/tokreducer/layers/semantic.py
"""Layer 3 — Semantic Compression.

Reduces prompts to semantic anchors — minimal identifiers the LLM expands
using its trained knowledge.  Applied at tok:3 (maximum compression) only.
"""

from __future__ import annotations

import re

_STOP_WORDS = re.compile(
    r"\b(?:the|a|an|of|for|is|it|its|this|that|these|those|"
    r"with|from|into|about|me|my|your|our|how|what|when|where|"
    r"which|who|whom|whose|here|there|"
    r"works?|provides?|gives?|makes?|takes?|gets?|has|have|had|"
    r"should|would|could|will|shall|may|might|must|"
    r"do|does|did|be|been|being|am|are|was|were|"
    r"all|each|every|some|any|no|not|"
    r"on|in|at|to|by|up|out|off|over|under|"
    r"full|or|if|as|and|so|yet|also|"
    r"please|explain|describe|write|create|generate|"
    r"following|main|key|important|relevant|specific|"
    r"include|including|section|detail|detailed|"
    r"commonly|currently|used|using|real-world|"
    r"different|designed|approaches|examples)\b",
    re.IGNORECASE,
)

_OPERATOR_PREFIX = re.compile(r"^[@>#!?~=•]|^(?:ctx:|fmt:|eg:|only>|len[=:]|tok:|\[)")

# Patterns to collapse multi-word phrases into single compact tokens.
_COMPOUND_REDUCTIONS: list[tuple[re.Pattern[str], str]] = [
    (re.compile(r"mechanism\s+(?:of\s+)?action", re.I), "mechanism"),
    (re.compile(r"drug\s+class(?:es)?", re.I), "drugs"),
    (re.compile(r"side\s+effects?", re.I), "effects"),
    (re.compile(r"treatment\s+protocols?", re.I), "protocols"),
    (re.compile(r"cancer\s+types?", re.I), "cancer"),
    (re.compile(r"clinical\s+trials?", re.I), "trials"),
    (re.compile(r"emerging\s+therap(?:y|ies)", re.I), "emerging-tx"),
    (re.compile(r"immunotherapy", re.I), "immuno"),
    (re.compile(r"medications?", re.I), "meds"),
    (re.compile(r"time[\s-]*complexity", re.I), "time-O"),
    (re.compile(r"space[\s-]*complexity", re.I), "space-O"),
    (re.compile(r"visual\s+(?:diagram\s+)?example", re.I), "visual"),
    (re.compile(r"renewable\s+energy", re.I), "renewables"),
    (re.compile(r"climate\s+change", re.I), "climate"),
    (re.compile(r"machine\s+learning", re.I), "ML"),
    (re.compile(r"artificial\s+intelligence", re.I), "AI"),
    (re.compile(r"natural\s+language\s+processing", re.I), "NLP"),
    (re.compile(r"deep\s+learning", re.I), "DL"),
    (re.compile(r"data\s+(?:set|analysis|science)", re.I), "data"),
    (re.compile(r"business\s+plan", re.I), "biz-plan"),
    (re.compile(r"marketing\s+strategy", re.I), "mktg"),
    (re.compile(r"software\s+engineer(?:ing)?", re.I), "sw-eng"),
    (re.compile(r"code\s+review", re.I), "code-review"),
]


class SemanticCompressor:
    """Layer 3: collapse prompts to semantic anchors at maximum compression."""

    def compress(self, text: str) -> str:
        result = text

        # Apply compound reductions first.
        for pattern, replacement in _COMPOUND_REDUCTIONS:
            result = pattern.sub(replacement, result)

        # Strip stop words.
        result = _STOP_WORDS.sub("", result)

        # Collapse whitespace around operators.
        result = re.sub(r"\s*\+\s*", "+", result)
        result = re.sub(r"\s*>>\s*", " >> ", result)

        # Keep only meaningful tokens (strip punctuation).
        tokens: list[str] = []
        for tok in result.split():
            tok = tok.strip(".,;:!?(){}\"'")
            if not tok or len(tok) <= 1:
                continue
            tokens.append(tok)

        result = " ".join(tokens)

        # Collapse adjacent plain words into "+" joined form.
        parts: list[str] = []
        buf: list[str] = []
        for tok in result.split():
            if _OPERATOR_PREFIX.match(tok) or "+" in tok or tok == ">>":
                if buf:
                    parts.append("+".join(w.lower() for w in buf))
                    buf = []
                parts.append(tok)
            else:
                buf.append(tok)
        if buf:
            parts.append("+".join(w.lower() for w in buf))

        result = " ".join(p for p in parts if p and p != "+")

        # Convert "explain" patterns to compact notation.
        result = re.sub(r"\bexplain[+:]([\w+-]+)", r"?\1", result)
        result = re.sub(r"\bexplain\b", "?", result)

        # Deduplicate >> operators.
        result = re.sub(r"(>>\s*)+", ">> ", result)

        # Global deduplication: collect all content words, keep only first
        # occurrence.  Operator-prefixed tokens are always kept.
        # At tok:3, also absorb standalone words that are sub-strings of
        # already-seen compound tokens (e.g. "meds" is redundant if "drugs"
        # is already present).
        seen: set[str] = set()
        deduped: list[str] = []
        for tok in result.split():
            if _OPERATOR_PREFIX.match(tok) or tok == ">>":
                deduped.append(tok)
                continue
            if "+" in tok:
                items = tok.split("+")
                unique = []
                for item in items:
                    low = item.strip().lower()
                    if low and low not in seen and not self._is_redundant(low, seen):
                        seen.add(low)
                        unique.append(item.strip())
                if unique:
                    deduped.append("+".join(unique))
            else:
                low = tok.strip().lower()
                if low and low not in seen and not self._is_redundant(low, seen):
                    seen.add(low)
                    deduped.append(tok)

        result = " ".join(deduped)

        # Final cleanup.
        result = re.sub(r"\+\+", "+", result)
        result = re.sub(r"  +", " ", result).strip()

        return result

    @staticmethod
    def _is_redundant(word: str, seen: set[str]) -> bool:
        """Check if *word* is semantically redundant given already-seen tokens.

        A word is redundant if it shares a root with an existing token
        or belongs to the same semantic cluster.
        """
        # Semantic clusters: words in the same cluster are interchangeable.
        _CLUSTERS: list[set[str]] = [
            {"drugs", "meds", "medications", "drug", "med", "medication", "pharmacology"},
            {"mechanism", "action", "moa", "how"},
            {"effects", "side-effects", "adverse", "toxicity", "risks"},
            {"protocols", "treatment", "therapy", "regimen", "trials", "clinical",
             "emerging-tx", "emerging", "novel", "new", "immuno", "immunotherapy", "immune"},
            {"cancer", "tumor", "oncology", "neoplasm", "malignancy"},
            {"complexity", "time-o", "space-o", "big-o", "performance"},
            {"visual", "diagram", "chart", "illustration", "figure"},
            {"steps", "step-by-step", "sequential", "procedure", "process"},
            {"review", "audit", "check", "inspect", "code-review"},
            {"summary", "summarize", "overview", "recap", "brief"},
        ]
        for cluster in _CLUSTERS:
            if word in cluster and seen & cluster:
                return True
        # Also check if word is a substring of any seen token.
        for s in seen:
            if len(word) >= 3 and (word in s or s in word):
                return True
        return False

    def decompress(self, text: str) -> str:
        """Semantic decompression is inherently lossy; return as-is."""
        return text

File: backend/rust/src/layers/semantic.rs
use std::collections::HashSet;
use regex::Regex;

static COMPOUND_REDUCTIONS: &[(&str, &str)] = &[
    (r"(?i)mechanism\s+(?:of\s+)?action", "mechanism"),
    (r"(?i)drug\s+class(?:es)?", "drugs"),
    (r"(?i)side\s+effects?", "effects"),
    (r"(?i)treatment\s+protocols?", "protocols"),
    (r"(?i)cancer\s+types?", "cancer"),
    (r"(?i)clinical\s+trials?", "trials"),
    (r"(?i)emerging\s+therap(?:y|ies)", "emerging-tx"),
    (r"(?i)immunotherapy", "immuno"),
    (r"(?i)medications?", "meds"),
    (r"(?i)time[\s-]*complexity", "time-O"),
    (r"(?i)space[\s-]*complexity", "space-O"),
    (r"(?i)visual\s+(?:diagram\s+)?example", "visual"),
    (r"(?i)renewable\s+energy", "renewables"),
    (r"(?i)climate\s+change", "climate"),
    (r"(?i)machine\s+learning", "ML"),
    (r"(?i)artificial\s+intelligence", "AI"),
    (r"(?i)business\s+plan", "biz-plan"),
    (r"(?i)software\s+engineer(?:ing)?", "sw-eng"),
    (r"(?i)code\s+review", "code-review"),
];

static SEMANTIC_CLUSTERS: &[&[&str]] = &[
    &["drugs", "meds", "medications", "drug", "med", "medication", "pharmacology"],
    &["mechanism", "action", "moa", "how"],
    &["effects", "side-effects", "adverse", "toxicity", "risks"],
    &["protocols", "treatment", "therapy", "regimen", "trials", "clinical",
      "emerging-tx", "emerging", "novel", "new", "immuno", "immunotherapy", "immune"],
    &["cancer", "tumor", "oncology", "neoplasm", "malignancy"],
    &["complexity", "time-o", "space-o", "big-o", "performance"],
    &["visual", "diagram", "chart", "illustration", "figure"],
    &["steps", "step-by-step", "sequential", "procedure", "process"],
    &["review", "audit", "check", "inspect", "code-review"],
    &["summary", "summarize", "overview", "recap", "brief"],
];

pub struct SemanticCompressor {
    stop_words: Regex,
    compound_patterns: Vec<(Regex, String)>,
    operator_prefix: Regex,
}

impl SemanticCompressor {
    pub fn new() -> Self {
        let stop_words = Regex::new(
            r"(?i)\b(?:the|a|an|of|for|is|it|its|this|that|these|those|with|from|into|about|me|my|your|our|how|what|when|where|which|who|whom|whose|here|there|works?|provides?|gives?|makes?|takes?|gets?|has|have|had|should|would|could|will|shall|may|might|must|do|does|did|be|been|being|am|are|was|were|all|each|every|some|any|no|not|on|in|at|to|by|up|out|off|over|under|full|or|if|as|and|so|yet|also|please|explain|describe|write|create|generate|following|main|key|important|relevant|specific|include|including|section|detail|detailed|commonly|currently|used|using|real-world|different|designed|approaches|examples)\b"
        ).unwrap();

        let compound_patterns = COMPOUND_REDUCTIONS
            .iter()
            .map(|(pat, repl)| (Regex::new(pat).unwrap(), repl.to_string()))
            .collect();

        let operator_prefix = Regex::new(r"^[@>#!?~=•]|^(?:ctx:|fmt:|eg:|only>|len[=:]|tok:|\[)").unwrap();

        Self { stop_words, compound_patterns, operator_prefix }
    }

    pub fn compress(&self, text: &str) -> String {
        let mut result = text.to_string();

        for (re, replacement) in &self.compound_patterns {
            result = re.replace_all(&result, replacement.as_str()).to_string();
        }

        result = self.stop_words.replace_all(&result, "").to_string();

        // Collapse whitespace around operators.
        let plus_ws = Regex::new(r"\s*\+\s*").unwrap();
        result = plus_ws.replace_all(&result, "+").to_string();
        let step_ws = Regex::new(r"\s*>>\s*").unwrap();
        result = step_ws.replace_all(&result, " >> ").to_string();

        // Keep only meaningful tokens.
        let tokens: Vec<String> = result
            .split_whitespace()
            .map(|t| t.trim_matches(|c: char| ".,;:!?(){}\"'".contains(c)).to_string())
            .filter(|t| !t.is_empty() && t.len() > 1)
            .collect();
        result = tokens.join(" ");

        // Collapse adjacent plain words into "+" joined form.
        let mut parts: Vec<String> = Vec::new();
        let mut buf: Vec<String> = Vec::new();
        for tok in result.split_whitespace() {
            if self.operator_prefix.is_match(tok) || tok.contains('+') || tok == ">>" {
                if !buf.is_empty() {
                    parts.push(buf.iter().map(|w| w.to_lowercase()).collect::<Vec<_>>().join("+"));
                    buf.clear();
                }
                parts.push(tok.to_string());
            } else {
                buf.push(tok.to_string());
            }
        }
        if !buf.is_empty() {
            parts.push(buf.iter().map(|w| w.to_lowercase()).collect::<Vec<_>>().join("+"));
        }
        result = parts.into_iter().filter(|p| !p.is_empty() && p != "+").collect::<Vec<_>>().join(" ");

        // Convert explain patterns.
        let explain_re = Regex::new(r"\bexplain[+:]([\w+-]+)").unwrap();
        result = explain_re.replace_all(&result, "?$1").to_string();
        let explain_bare = Regex::new(r"\bexplain\b").unwrap();
        result = explain_bare.replace_all(&result, "?").to_string();

        // Deduplicate >> operators.
        let dup_step = Regex::new(r"(>>\s*)+").unwrap();
        result = dup_step.replace_all(&result, ">> ").to_string();

        // Global deduplication with semantic clusters.
        let mut seen: HashSet<String> = HashSet::new();
        let mut deduped: Vec<String> = Vec::new();
        for tok in result.split_whitespace() {
            if self.operator_prefix.is_match(tok) || tok == ">>" {
                deduped.push(tok.to_string());
                continue;
            }
            if tok.contains('+') {
                let items: Vec<&str> = tok.split('+').collect();
                let unique: Vec<&str> = items
                    .into_iter()
                    .filter(|item| {
                        let low = item.trim().to_lowercase();
                        if low.is_empty() || seen.contains(&low) || Self::is_redundant(&low, &seen) {
                            false
                        } else {
                            seen.insert(low);
                            true
                        }
                    })
                    .collect();
                if !unique.is_empty() {
                    deduped.push(unique.join("+"));
                }
            } else {
                let low = tok.trim().to_lowercase();
                if !low.is_empty() && !seen.contains(&low) && !Self::is_redundant(&low, &seen) {
                    seen.insert(low);
                    deduped.push(tok.to_string());
                }
            }
        }
        result = deduped.join(" ");

        // Final cleanup.
        let dup_plus = Regex::new(r"\+\+").unwrap();
        result = dup_plus.replace_all(&result, "+").to_string();
        let dup_space = Regex::new(r"  +").unwrap();
        result = dup_space.replace_all(&result, " ").trim().to_string();

        result
    }

    fn is_redundant(word: &str, seen: &HashSet<String>) -> bool {
        for cluster in SEMANTIC_CLUSTERS {
            if cluster.contains(&word) && cluster.iter().any(|c| seen.contains(*c)) {
                return true;
            }
        }
        for s in seen {
            if word.len() >= 3 && (word.contains(s.as_str()) || s.contains(word)) {
                return true;
            }
        }
        false
    }

    pub fn decompress(&self, text: &str) -> String {
        text.to_string()
    }
}

impl Default for SemanticCompressor {
    fn default() -> Self {
        Self::new()
    }
}
